src/Owin.Security.OpenIdConnect.Server/OpenIdConnectServerHandler.Endpoints.cs

```diff
@@ -1625,97 +1649,160 @@ internal partial class OpenIdConnectServerHandler : AuthenticationHandler<OpenId
+                    }
+                    
+                    payload.Add(OpenIdConnectConstants.Claims.ClientId, request.ClientId);
 
-            payload.Add("audiences", JArray.FromObject(ticket.Properties.GetAudiences()));
-            payload.Add("expires_in", ticket.Properties.ExpiresUtc.Value);
-            
-            payload.Add("claims", JArray.FromObject(
-                from claim in notification.Claims
-                select new { type = claim.Type, value = claim.Value }
-            ));
+                    // TODO: Re-examine whether ticket.Identity will always be the resource owner who authorized the token
+                    // and whether Username should always be returned
+                    if (!string.IsNullOrWhiteSpace(ticket.Identity.Name)) {
+                        payload.Add(OpenIdConnectConstants.Claims.Username, ticket.Identity.Name);
+                    }
 
-            var context = new ValidationEndpointResponseContext(Context, Options, payload);
-            await Options.Provider.ValidationEndpointResponse(context);
+                    // TODO: Determine the token type for the given token when not an access token
+                    // See https://tools.ietf.org/html/rfc7662#section-2.2 and https://tools.ietf.org/html/rfc6749#section-5.1
+                    if (string.Equals(ticket.GetUsage(), OpenIdConnectConstants.Usages.AccessToken)) {
+                        payload.Add(OpenIdConnectConstants.Claims.TokenType, "Bearer");
+                    }
 
-            if (context.HandledResponse) {
+                    if (ticket.Properties.ExpiresUtc.HasValue) {
+                        payload.Add(JwtRegisteredClaimNames.Exp, ticket.Properties.ExpiresUtc.Value.Ticks);
+                    }
+
+                    // TODO: Determine if this implementation always issues tokens that are active upon issuance, or if the active time can start later
+                    // If not, iat and nbf could be different values and should be accounted for separately.
+                    if (ticket.Properties.IssuedUtc.HasValue) {
+                        payload.Add(JwtRegisteredClaimNames.Iat,
+                            EpochTime.GetIntDate(ticket.Properties.IssuedUtc.Value.UtcDateTime).ToString());
+                        payload.Add(JwtRegisteredClaimNames.Nbf,
+                            EpochTime.GetIntDate(ticket.Properties.IssuedUtc.Value.UtcDateTime).ToString());
+                    }
+
+                    // Note: uses the name claim type in the identity instead of the specific ClaimTypes.NameIdentifier to get the subject
+                    // TODO: Determine if this should be the correct logic for the middleware instead of being tightly-coupled for the name claim type
+                    var subject = ticket.Identity.GetClaim(JwtRegisteredClaimNames.Sub) ??
+                                  ticket.Identity.GetClaim(ticket.Identity.NameClaimType);
```

Comment:
Ah, I see. I did not know the reason behind using `ClaimTypes.NameIdentifier` elsewhere. That makes sense, though doesn't `Identity.NameClaimType` have to be unique as well for Identity? Pretty sure it does for a `SecurityStampValidator` but perhaps I'm mistaken...